### ISSUE TITLE
('Connection aborted.', OSError("(104, 'ECONNRESET')"))

### ISSUE THREAD

**Issue by Moebius895** [eternnoir/pyTelegramBotAPI#880]
1. What version of pyTelegramBotAPI are you using?
Latest

2. What OS are you using?
Xubuntu 20.04

3. What version of python are you using?
3.8

It seems to me that this problem is pretty common, bur since I could not find any solution, I would like to submit it again.
The bot I am building is currently hosted on a linux virtual machine on my computer, and uses polling to get updates.

The bot reads all messages in a group, so that it can keep track of most and least active members.
But I also implemented functions to handle user input, such as commands or buttons. The problem is, if no user gives any input for a sufficient amount of time (eg. 10-15 minutes), the next time somebody does, the request to the relative Telegram API will raise
`('Connection aborted.', OSError("(104, 'ECONNRESET')"))`

so that the bot has to start a new connection, and the user has to wait for between 30 and 60 seconds to see the expected output.

Here is a simplified version of my code (it could be used to reproduce the problem)

```
import telebot
import logging
import schedule
import time
from threading import Thread

TOKEN = TELEGRAM_BOT_TOKEN
bot = telebot.TeleBot(TOKEN)

logger = telebot.logger
log_format='%(asctime)s - %(levelname)s - %(message)s'
logging.basicConfig(format=log_format,level=logging.DEBUG,datefmt='%-d %b %Y - %H:%M:%S')


group_id = GROUP_ID       #The group where my bot counts messages


message_counter = {}

@bot.message_handler(commands=['start','help'])                 #Handles commands /start and /help
def start_fun(message):
    user_id = message.from_user.id
    logging.info(str(user_id)+" requested /start or /help.")
    for i in range(10):                                         #Maximum of 10 retries
        try:
            logging.info("Attempt no. "+str(i+1))
            bot.send_message(message.chat.id,"Hello!")
        except Exception as e:
            logging.error(e)
        else:
            logging.info("Done")
            break

@bot.message_handler(func = lambda m : True)                     #Handles every message, inside the group
def general_handler(message):                                    #Passively counts the messages for each 
    if (message.chat.id != group_id):                            #user, only inside the group
        return
    user_id = str(message.from_user.id)
    logging.info("Message from "+user_id)
    if (user_id not in message_count):
        message_count[user_id] = 0
    else:
        message_count[user_id] += 1


def scheduled_task():                                             #example of scheduled task
    for i in range(10):
        try:
            logging.info("Trying to do scheduled task")
            bot.send_message(group_id,"I will say this every day!")
        except Exception as e:
            logging.error(e)
        else:
            logging.info("Done")


def schedule_checker():                                          #checks every seconds for pending tasks
    while True:
         schedule.run_pending()
         time.sleep(1)

schedule.every().day.at("13:00").do(scheduled_task)
Thread(target=schedule_checker).start()                           #Opens a new thread so it does not
                                                                                          #interfere with polling


while True:                                                       #Loop with infinite retries, so it
        try:                                                         #does not stop if there is an error
                bot.polling(none_stop=True)
        except Exception as e:
                logging.error(e)
                time.sleep(1)

```

This is an example from my debug log

```
15 Jun 2020 - 10:48:32 - DEBUG - Starting new HTTPS connection (1): api.telegram.org:443
15 Jun 2020 - 17:47:00 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=1&timeout=20 HTTP/1.1" 200 23
15 Jun 2020 - 17:47:20 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=1&timeout=20 HTTP/1.1" 200 23
15 Jun 2020 - 17:47:40 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=1&timeout=20 HTTP/1.1" 200 23
15 Jun 2020 - 17:48:00 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=1&timeout=20 HTTP/1.1" 200 23
15 Jun 2020 - 17:48:20 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=1&timeout=20 HTTP/1.1" 200 23
15 Jun 2020 - 17:50:07 - INFO - Message from 123456789
15 Jun 2020 - 17:50:16 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861246&timeout=20 HTTP/1.1" 200 284
15 Jun 2020 - 17:50:16 - INFO - Message from 234567891
15 Jun 2020 - 17:50:26 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861247&timeout=20 HTTP/1.1" 200 303
15 Jun 2020 - 17:50:26 - INFO - Message from 234567891
15 Jun 2020 - 17:50:42 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861248&timeout=20 HTTP/1.1" 200 562
15 Jun 2020 - 17:50:42 - INFO - Message from 123456789
15 Jun 2020 - 17:50:44 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861249&timeout=20 HTTP/1.1" 200 318
15 Jun 2020 - 17:50:44 - INFO - Message from 123456789
15 Jun 2020 - 17:50:51 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861250&timeout=20 HTTP/1.1" 200 292
15 Jun 2020 - 17:50:51 - INFO - Message from 234567891
15 Jun 2020 - 17:51:07 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861251&timeout=20 HTTP/1.1" 200 324
15 Jun 2020 - 17:51:07 - INFO - Message from 123456789
15 Jun 2020 - 17:51:10 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861252&timeout=20 HTTP/1.1" 200 553
15 Jun 2020 - 17:51:10 - INFO - Message from 123456789
15 Jun 2020 - 19:20:21 - INFO - 11223344 requested /start or /help
15 Jun 2020 - 19:20:21 - INFO - Attempt no. 1
15 Jun 2020 - 19:20:40 - ERROR - ('Connection aborted.', OSError("(104, 'ECONNRESET')"))
15 Jun 2020 - 19:20:41 - DEBUG - https://api.telegram.org:443 "GET /TOKEN/getUpdates?offset=456861489&timeout=20 HTTP/1.1" 200 23
15 Jun 2020 - 19:20:41 - INFO - Attempt no. 2
15 Jun 2020 - 19:20:41 - DEBUG - Starting new HTTPS connection (4): api.telegram.org:443
15 Jun 2020 - 19:20:41 - DEBUG - https://api.telegram.org:443 "POST /TOKEN/sendMessage?chat_id=11223344&text=Hello!" 200 2878
15 Jun 2020 - 19:20:41 - INFO - Done
```

To sum up:

1) No problems as long as the get_updates() function is called to retrieve updates, such as for counting messages in the group

2) When a command is given to the bot, it might give the error and then take 20-30 seconds to start a new connection (the second attempt is always successful)

3) The same problem happens when trying to execute scheduled tasks (which, in my case, all require calling the Telegram API functions)

4) The connection reset error does not happen if the commands are given in a short amount of time, i.e. the error will happen again if no command is given in 10-15 minutes

Thanks to the try-except loops I managed to prevent the bot from crashing, but users still often face these long waiting times. 

Is there a solution?

Thanks

**Comment by karmaBonfire** [comment]
I have the same problem. After a period of inactivity (~15 min.) the update is received normally, but an outgoing request like sending a message (a voice message in my case) fails either with `ConnectionResetError` or `ReadTimeout`.

pyTelegramBotAPI: 3.7.1
OS: Ubuntu 18.04
Python: 3.6

**Comment by raitonoberu** [comment]
The only solution I found:

    import telebot
    from requests.exceptions import ReadTimeout

    telebot.apihelper.READ_TIMEOUT = 5  # you can also set a lower value

    def send_msg(id, text):
        try:
            bot.send_message(id, text, reply_markup=keyboard)
        except ReadTimeout:
            send_msg(text, id)

**Comment by Moebius895** [comment]
> The only solution I found:
> 
> ```
> from requests.exceptions import ReadTimeout
> 
> telebot.apihelper.READ_TIMEOUT = 5  # you can also set a lower value
> 
> def send_msg(id, text):
>     try:
>         bot.send_message(id, text)
>     except ReadTimeout:
>         send_msg(text, id)
> ```

Already tried to reduce the read timeout. This reduces the waiting times in case of a `ReadTimeout `error, but does not help me with the `ConnectionResetError`

**Comment by raitonoberu** [comment]
I use this code and I have no connection issues at all:

    import telebot
    from time inport sleep
    import logging

    telebot.apihelper.READ_TIMEOUT = 5
    
    bot = telebot.TeleBot(TOKEN, threaded=False)
    # "threaded=False" is much more stable

    def send_msg(id, text):
        try:
            bot.send_message(id, text)
        except Exception as e:
            logging.error(e)
            sleep(1)
            send_msg(id, text)
    
    ### Your code ####
    
    bot.infinity_polling(none_stop=True)  # use this instead of bot.pooling()

**Comment by Moebius895** [comment]
> I use this code and I have no connection issues at all:
> 
> ```
> import telebot
> from time inport sleep
> import logging
> 
> telebot.apihelper.READ_TIMEOUT = 5
> 
> bot = telebot.TeleBot(TOKEN, threaded=False)
> # "threaded=False" is much more stable
> 
> def send_msg(id, text):
>     try:
>         bot.send_message(id, text)
>     except Exception as e:
>         logging.error(e)
>         sleep(1)
>         send_msg(id, text)
> 
> ### Your code ####
> 
> bot.infinity_polling(none_stop=True)  # use this instead of bot.pooling()
> ```

Thank you very much, I have not seen the error anymore after adding `threaded = False`. I'll close the issue as soon as I am sure that the problem is permanently solved
